test_quality_of_service/test/test_deadline.cpp

```
@@ -0,0 +1,121 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <tuple>
+
+#include "gtest/gtest.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/single_threaded_executor.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "test_quality_of_service/qos_utilities.hpp"
+
+
+using namespace std::chrono_literals;
+
+/// Test Deadline with a single publishing node and single subscriber node
+TEST_F(TestSetup, test_deadline) {
+  int expected_number_of_pub_events = 5;
+  const std::chrono::milliseconds deadline_duration = 1s;
+  std::tuple<size_t, size_t> deadline_duration_tuple = chrono_milliseconds_to_size_t(
+    deadline_duration);
+  const std::chrono::milliseconds max_test_length = 10s;
+  const std::chrono::milliseconds publish_rate = deadline_duration / 5;
+
+  // used for lambda capture
+  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec = executor;
+  int total_number_of_publisher_deadline_events = 0;
+  int total_number_of_subscriber_deadline_events = 0;
+  int last_pub_count = 0;
+  int last_sub_count = 0;
+
+  rmw_qos_profile_t qos_profile = rmw_qos_profile_default;
+  qos_profile.deadline.sec = std::get<0>(deadline_duration_tuple);
+  qos_profile.deadline.nsec = std::get<1>(deadline_duration_tuple);
+
+  // setup subscription options and callback
+  rclcpp::SubscriptionOptions<> subscriber_options;
+  subscriber_options.qos_profile = qos_profile;
+  subscriber_options.event_callbacks.deadline_callback =
+    [&last_pub_count,
+    &total_number_of_subscriber_deadline_events](rclcpp::QOSDeadlineRequestedInfo & event) -> void
+    {
+      RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "QOSDeadlineRequestedInfo callback");
+      total_number_of_subscriber_deadline_events++;
+      // assert the correct value on a change
+      ASSERT_EQ(1, event.total_count_change);
```

EXPECT? https://testing.googleblog.com/2008/07/tott-expect-vs-assert.html

---

Depending on how rclcpp is implemented, asserting there may make the whole test crash which is annoying. Switching a boolean we'd check in the main thread is a way to avoid this problem if it happens.

---

> EXPECT? https://testing.googleblog.com/2008/07/tott-expect-vs-assert.html

Assert was used explicitly. If the callback fails there's no reason to burn CPU cycles (at the end of the test experiment) when we know that the rest of the test code will fail. The intent here was simply to fail fast.